examples_tests/22.RaytracedAO/raytraceCommon.glsl

```diff
@@ -252,14 +260,82 @@ vec3 load_normal_and_prefetch_textures(
 }
 
+vec3 nbl_glsl_unormSphericalToCartesian(in vec2 uv, out float sinTheta)
+{
+	vec3 dir;
+	nbl_glsl_sincos((uv.x-0.5)*2.f*nbl_glsl_PI,dir.y,dir.x);
+	nbl_glsl_sincos(uv.y*nbl_glsl_PI,sinTheta,dir.z);
+	dir.xy *= sinTheta;
+	return dir;
+}
+
+// return regularized pdf of sample
+float Envmap_regularized_deferred_pdf(in vec3 rayDirection)
+{
+	const ivec2 luminanceMapSize = textureSize(luminance[0], 0);
+	uint lastLuminanceMip = uint(log2(luminanceMapSize.x)); // TODO: later turn into push constant
+	const vec2 envmapUV = nbl_glsl_sampling_generateUVCoordFromDirection(rayDirection);
+
+	float sinTheta = length(rayDirection.zx);
+	float sumLum = texelFetch(luminance[lastLuminanceMip], ivec2(0), 0).r;
+	float lum = textureLod(luminance[0], envmapUV, 0).r;
+	float bigfactor = float(luminanceMapSize.x*luminanceMapSize.y)/sumLum;
+	return bigfactor*(lum/(sinTheta*2.0f*nbl_glsl_PI*nbl_glsl_PI));
+}
+
+void Envmap_generateRegularizedSample_and_pdf(out float pdf, out nbl_glsl_LightSample lightSample, in nbl_glsl_AnisotropicViewSurfaceInteraction interaction, in vec2 rand)
+{
+	const ivec2 warpMapSize = textureSize(warpMap,0);
+	const ivec2 lastWarpMapPixel = warpMapSize - ivec2(1.f);
+	vec2 xi = rand;
+	const vec2 unnormCoord = xi*lastWarpMapPixel;
+
+	const vec2 interpolant = fract(unnormCoord);
+
+	vec2 warpSampleCoord = (unnormCoord+vec2(0.5f))/vec2(warpMapSize);
+	const vec4 dirsX = textureGather(warpMap, warpSampleCoord, 0); // 0_1, 1_1, 1_0, 0_0
+	const vec4 dirsY = textureGather(warpMap, warpSampleCoord, 1); // 0_1, 1_1, 1_0, 0_0
+	const mat4x2 uvs = transpose(mat2x4(dirsX,dirsY));
+
+	const vec2 xDiffs[] = {
+		uvs[2]-uvs[3],
+		uvs[1]-uvs[0]
+	};
+	const vec2 yVals[] = {
+		xDiffs[0]*interpolant.x+uvs[3],
+		xDiffs[1]*interpolant.x+uvs[0]
+	};
+	const vec2 yDiff = yVals[1]-yVals[0];
+	const vec2 uv = yDiff*interpolant.y+yVals[0];
+
+	float sinTheta;
+	const vec3 L = nbl_glsl_unormSphericalToCartesian(uv, sinTheta);
+	lightSample = nbl_glsl_createLightSample(L, interaction);
+	
+	const float detInterpolJacobian = determinant(mat2(
+		mix(xDiffs[0],xDiffs[1],interpolant.y), // first column dFdx
+		yDiff // second column dFdy
+	));
+
+	float pdfConstant = 1.f/(4.f*nbl_glsl_PI*float(lastWarpMapPixel.x*lastWarpMapPixel.y));
+	pdf = pdfConstant/abs(sinTheta*detInterpolJacobian);
+}
+
 #include <nbl/builtin/glsl/sampling/quantized_sequence.glsl>
-vec3 rand3d(in uvec3 scramble_key, in int _sample, int depth)
+mat2x3 rand6d(in uvec3 scramble_key, in int _sample, int depth)
 {
+	mat2x3 retVal;
 	// decrement depth because first vertex is rasterized and picked with a different sample sequence
 	--depth;
 	//
-	const nbl_glsl_sampling_quantized3D quant = texelFetch(quantizedSampleSequence,int(_sample)*SAMPLE_SEQUENCE_STRIDE+depth).xy;
-    return nbl_glsl_sampling_decodeSample3Dimensions(quant,scramble_key);
+	int offset = int(_sample)*SAMPLE_SEQUENCE_STRIDE+depth;
+	int eachStrategyStride = SAMPLE_SEQUENCE_STRIDE/2; // get this from cpp side?
```

Comment:
```glsl
const int eachStrategyStride = SAMPLE_SEQUENCE_STRIDE/SAMPLING_STRATEGY_COUNT;
```
should be decent and understandable enough.

Reply:
Done